# Data analysis of TMDB
Ih this task I am going to do some data analysis of movies and tv shows that are in The Movie DataBase.
The insights I am going to find is the following:
- What directors have made most movies and/or tv-shows.
- What actors have been the most active.
- What genres was the most popular per decade.

## Getting the data

I was at first going to just use [TMDB's API](https://developer.themoviedb.org/reference/intro/getting-started), but I encountered a problem. When I was trying to call for the 501th page (i.e. pages 500+) it gave me a error with the text `Invalid page: Pages start at 1 and max at 500. They are expected to be an integer.`, this is quite annoying since the amount of pages as today (16 01 2025) is a whoping `48184`. And with the page limit I would only be able to access **just a bit over 1%** using this method. I did then discover the [daily ID](https://developer.themoviedb.org/docs/daily-id-exports) lists that they have. This was pretty sweet, but sadly i stumbled upon a me problem this time, my code was not good enough and I did not know how to improve it further, a aproximatly 190 hours of processing time to find directors from just the `People` list? Not acceptable, I then went for something a bit easier, I found this lovely [.csv file](https://www.kaggle.com/datasets/alanvourch/tmdb-movies-daily-updates) that user Alan Vourc'h had uploaded to Kaggle, and that file is what I am goign to use for this project.

## Start of code
I am going to use Python as my coding language, since it is what I am the most comfortable with.

The very first thing I do is choose my modules for my code:
- I chose `pandas` as my module of choice to handle the .csv file.
- I chose `numpy` for its usefull math library.
- I chose `seaborn` for its posibility to visualize data in a friendly way.
- I chose `matplotlib.pyplot` for making graphing possible.

In [14]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


Then since the file itself has many columns that I do not need, I made a list for them, and put that list in to the `usecols` parametre in panas' `read_csv`.

In [15]:

fields = ["title","id","vote_average","vote_count","release_date","revenue","budget","runtime","genres","cast","director","writers","producers","imdb_rating","imdb_votes"] # Put the info you want here

df = pd.read_csv(r"Data\TMDB_all_movies.csv", skipinitialspace=True, usecols=fields)

Then, since there are some rows that had `NAN` as their values, i changed those to a 0 where it was fitting. (I did not do this for the movies with no credits, they are filtered out.) 
I also set the index to be the movies ID.

In [16]:
#Data cleaning START
df.set_index("id", inplace=True) #Sets the index to be the movies ID

for i in ["vote_average","vote_count","revenue","budget","imdb_rating","imdb_votes"]:
    df[i].fillna(0,inplace=True)
    #This replaces NAN values with in the list to 0

df["release_date"] = pd.to_datetime(df["release_date"]) #Turns the dates in to the date format

df.drop_duplicates(inplace = True)

df.dropna(inplace=True)
#Data cleaning END

C:\Users\Bruker\AppData\Local\Temp\ipykernel_6924\2791509396.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[i].fillna(0,inplace=True)


## The First Question

So to find out who has directed the most movies, and how many movies they have directed I do a quick query on the `director` collumn.

In [17]:
df["director"].describe(include="object")

count             227046
unique            107291
top       David DeCoteau
freq                 142
Name: director, dtype: object

And here I look at the `top` in the terminal to find out that it is David DeCoteau that has directed the most movies, and by looking at `freq` that he has directed 142 movies.

## The Second Question

Now to find the person whom has acted in the most movies, this is found in a very similar way to the first question, but now I will use the collumn `cast`.

In [18]:
df["cast"].describe(include="object")

count        227046
unique       224458
top       Mel Blanc
freq            139
Name: cast, dtype: object

And it is Mel Blanc that has played in the most movies, in 139 of them.

## The Third Question

Now to find out what genres are the most popular per decade, I first divide up each movie in to the decades that they are in.

In [21]:
df["decade"] = (df["release_date"].dt.year // 10) * 10

title  vote_average  vote_count release_date     revenue  \
id                                                                           
2                 Ariel         7.100       338.0   1988-10-21         0.0   
3   Shadows in Paradise         7.300       402.0   1986-10-17         0.0   
5            Four Rooms         5.900      2650.0   1995-12-09   4257354.0   
6        Judgment Night         6.500       333.0   1993-10-15  12136938.0   
9      Sunday in August         7.135        26.0   2004-09-02         0.0   

    runtime      budget                         genres  \
id                                                       
2      73.0         0.0  Comedy, Drama, Romance, Crime   
3      74.0         0.0         Comedy, Drama, Romance   
5      98.0   4000000.0                         Comedy   
6     109.0  21000000.0        Action, Crime, Thriller   
9      15.0         0.0                          Drama   

                                                 cast  \
id                                                      
2   Jyrki Olsonen, Jaakko Talaskivi, Olli Varja, H...   
3   Sirkka Silin, Helmeri Pellonpää, Tanja Talaski...   
5   Lili Taylor, Valeria Golino, Marisa Tomei, Mar...   
6   Stephen Dorff, Everlast, Will Zahrn, Emilio Es...   
9                          Rita Lengyel, Milton Welsh   

                                             director  \
id                                                      
2                                      Aki Kaurismäki   
3                                      Aki Kaurismäki   
5   Quentin Tarantino, Robert Rodriguez, Allison A...   
6                                     Stephen Hopkins   
9                               Marc Meyer, Anna Haas   

                                              writers  \
id                                                      
2                                      Aki Kaurismäki   
3                                      Aki Kaurismäki   
5   Quentin Tarantino, Robert Rodriguez, Allison A...   
6                       Jere Cunningham, Lewis Colick   
9                                          Marc Meyer   

                                            producers  imdb_rating  \
id                                                                   
2                                      Aki Kaurismäki          7.4   
3                                     Mika Kaurismäki          7.5   
5   Quentin Tarantino, Lawrence Bender, Alexandre ...          6.7   
6               Gene Levy, Marilyn Vance, Lloyd Segan          6.6   
9                                          Marc Meyer          6.8   

    imdb_votes  decade  
id                      
2       8892.0    1980  
3       7681.0    1980  
5     113177.0    1990  
6      19493.0    1990  
9         14.0    2000

I then run a `df.head()` to make sure that the new collumn `decade` has been added.

In [ ]:
df.head()

And it has!